### PR TITLE
Complete Ch4

### PR DESCRIPTION
**Pull request by rootbid**
### Solutions for Chapter 4

cc @vrom911 @chshersh

Here's my shot at Functors, Applicatives & Monads. Fascinating stuff 👍🏼

### CODE REVIEW

File: src/Chapter4.hs [hunk 1]
```diff
@@ -472,10 +487,11 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    (<*>) (Trap e) _ = Trap e
+    (<*>) (Reward f) x = fmap f x
```
Comment (chshersh) [inline src/Chapter4.hs:494]: Nice work on reusing `Functor` instance for `Secret`! :world_map:

File: src/Chapter4.hs [hunk 2]
```diff
@@ -610,6 +642,14 @@ Implement the 'Monad' instance for our lists.
 🕯 HINT: You probably will need to implement a helper function (or
   maybe a few) to flatten lists of lists to a single list.
 -}
+join :: List (List a) -> List a
+join Empty       = Empty
+join (Cons x xs) = x `append` (join xs)
+
+instance Monad List where
+    (>>=) :: List a -> (a -> List b) -> List b
+    (>>=) Empty _ = Empty
```
Comment (chshersh) [inline src/Chapter4.hs:651]: This case is redundant, because both `join` and `fmap` return `Empty` on `Empty` list :slightly_smiling_face:

**Review by chshersh — APPROVED** [review]
:tada: Fantastic work!
Thank you so much for taking this course! I hope you had fun :man_dancing: :microphone: 

We would really appreciate your [feedback on the course](https://docs.google.com/forms/d/e/1FAIpQLScBVhLxq5CgGnAfIGUE-fCoOUqeGkDY2HXzbT7KV2jjLOsmjQ/viewform)!

You can also ask any questions and we can discuss any topics you would like to dig into as well  :slightly_smiling_face: Don't be a stranger!

**Comment by rootbid** [comment]
I appreciate your feedback @chshersh .
This course was both fun and interesting. Kudos to your efforts! 👏🏼 

Will like to improve upon my understanding of Haskell type system and dig myself deeper into why some haskell developers hate working with I/O. 😄 

Also, I would very much like to contribute in open source haskell projects if you could give me some advice on where to start contributing or maybe continue my haskell learning adventure?

**Comment by chshersh** [comment]
@rootbid Don't be afraid of IO! It's not that scary, and there's nothing to hate actually :smile: 

If you want to continue learning Haskell, here is the list of maintainers you can check out, there could be some interesting projects for you: https://github.com/willbasky/Awesome-list-of-Haskell-mentors

Also, you could check the projects that are in Hacktoberfest, they usually have some issues you could tackle :slightly_smiling_face: 

Of course, we also have a lot of open-source Haskell projects. You can check them out and see if there's something interesting to contribute to :slightly_smiling_face:

**Pull request merged**